Add Clerk Express SDK docs (#1201)
clerk/clerk-docs: docs/quickstarts/express.mdx

    1. Paste your keys into your `.env.local` file.

    The final result should resemble the following:
  </SignedOut>

  ```sh {{ filename: '.env.local' }}
  CLERK_PUBLISHABLE_KEY={{pub_key}}
  CLERK_SECRET_KEY={{secret}}
  ```

  This examples uses `dotenv` to load the environment variables. You can use any other library or method, if you so wish.

  <CodeBlockTabs type="installer" options={["npm", "yarn", "pnpm"]}>
    ```bash {{ filename: 'terminal' }}
    npm install dotenv
    npm install -D @types/dotenv
    ```

    ```bash {{ filename: 'terminal' }}
    yarn add dotenv
    yarn add -D @types/dotenv
    ```

    ```bash {{ filename: 'terminal' }}
    pnpm add dotenv
    pnpm add -D @types/dotenv
    ```
  </CodeBlockTabs>

  ### Add middleware to your application

  The `clerkMiddleware` can be registered globally or for specific routes. This examples registers the middleware globally.

  ```ts {{ filename: 'index.ts' }}
  import 'dotenv/config'
  import express from 'express'
  import { clerkMiddleware } from '@clerk/express'

  const app = express()

  app.use(clerkMiddleware())

  app.listen(3000, () => {

>     console.log(`Example app listening at http://localhost:${port}`)

Reviewer: so is port set in the environment variables ?

Reviewer 2: Good catch, updated!

  })
  ```

>   ### Accessing auth state using `getAuth`

Reviewer: ```suggestion
  ### Accessing auth state using `getAuth()`
```

Reviewer 2: Updated this one, didn't know you started reviewing it already. My bad!

>   The [`getAuth`](/docs/references/nextjs/get-auth) helper can be used to access the authentication state of the current request. It can also be used to protect your routes.

Reviewer: We have to explain all the moving parts here

```suggestion
  The [`getAuth()`](/docs/references/express/overview#get-auth) helper retrieves authentication state from the request object.

  [`clerkClient`](/docs/references/express/overview#clerk-client) is an instance of Clerk's JavaScript Backend SDK, which provides access to Clerk's Backend API resources and low-level authentication utilities for JavaScript environments.

  The following example uses `getAuth()` to retrieve the `userId`, which is used to protect the route and is passed to `clerkClient.users.getUser()` to retrieve the current user's `User` object.
```

Reviewer 2: Updated!

>   ```ts {{ filename: 'index.ts' }}

Reviewer: I don't know if you tested this example, because `clerkClient` isn't being imported.

Reviewer 2: Updated this one!

  import 'dotenv/config'
  import express from 'express'
  import { clerkMiddleware, getAuth } from '@clerk/express'

  const app = express()

  app.use(clerkMiddleware())

  app.get('/protected', async (req, res) => {
    /**
     * Access the auth state for this request.
     * In this example, the userId loads the whole User object
     * from the Clerk servers
     */
    const { userId } = getAuth(req)

    if (!userId) {
      return res.status(401).send('Unauthorized')
    }

    const user = await clerkClient.users.getUser(userId)

    return res.json({ user })
  })

  app.listen(3000, () => {

>     console.log(`Example app listening at http://localhost:${port}`)

Reviewer: same question here about port

Reviewer 2: Updated! Thank you

---
title: Express Quickstart
description: Learn how to use Clerk to quickly and easily add secure authentication and user management to your Express server.
---

<TutorialHero
  framework="express"
  exampleRepo={[
    {
      title: "Express Quickstart Repo",
      link: "https://github.com/clerk/clerk-express-quickstart"
    }
  ]}
  beforeYouStart={[
    {
      title: "Set up a Clerk application",
      link: "/docs/quickstarts/setup-clerk",
      icon: "clerk",
    },
    {
      title: "Create a Express application",
      link: "https://expressjs.com/en/starter/installing.html",
      icon: "expressjs",
    }
  ]}
>
  - Install `@clerk/express`
  - Set your Clerk API keys

>   - Add Clerk middleware to your application

Reviewer: ```suggestion
  - Add `clerkMiddleware()` to your application
```

  - Access auth state using `getAuth()`
</TutorialHero>

Learn how to integrate Clerk into your Express backend for secure user authentication and management. This guide uses TypeScript and allows you to choose your frontend framework.

<Steps>
  ### Install `@clerk/express`

>   Add Clerk's Express SDK to your project:

Reviewer: ```suggestion
  Clerk's [Express SDK](/docs/references/express/overview) has prebuilt components, React hooks, and helpers to make user authentication easier.

To get started using Clerk with Express, add the SDK to your project:
```

  <CodeBlockTabs type="installer" options={["npm", "yarn", "pnpm"]}>
    ```bash {{ filename: 'terminal' }}
    npm install @clerk/express
    ```

    ```bash {{ filename: 'terminal' }}
    yarn add @clerk/express
    ```

    ```bash {{ filename: 'terminal' }}
    pnpm add @clerk/express
    ```
  </CodeBlockTabs>

  ### Set your Clerk API keys

  <SignedIn>
    Add the following keys to your `.env.local` file. These keys can always be retrieved from the [API Keys](https://dashboard.clerk.com/last-active?path=api-keys) page of your Clerk Dashboard.
  </SignedIn>

  <SignedOut>
    1. Navigate to the Clerk Dashboard.
    1. In the navigation sidebar, select [API Keys](https://dashboard.clerk.com/last-active?path=api-keys).
    1. In the **Quick Copy** section, copy your Clerk publishable and secret key.
    1. Paste your keys into your `.env.local` file.

    The final result should resemble the following:
  </SignedOut>

  ```sh {{ filename: '.env.local' }}
  CLERK_PUBLISHABLE_KEY={{pub_key}}
  CLERK_SECRET_KEY={{secret}}
  ```

>   This examples uses `dotenv` to load the environment variables. You can use any other library or method, if you so wish.

Reviewer: ```suggestion
  This guide uses `dotenv` to load the environment variables. You can use any other library or method, if you so wish.
```

  <CodeBlockTabs type="installer" options={["npm", "yarn", "pnpm"]}>
    ```bash {{ filename: 'terminal' }}
    npm install dotenv
    npm install -D @types/dotenv
    ```

    ```bash {{ filename: 'terminal' }}
    yarn add dotenv
    yarn add -D @types/dotenv
    ```

    ```bash {{ filename: 'terminal' }}
    pnpm add dotenv
    pnpm add -D @types/dotenv
    ```
  </CodeBlockTabs>

>   ### Add middleware to your application
> 
>   The `clerkMiddleware` can be registered globally or for specific routes. This examples registers the middleware globally.

Reviewer: ```suggestion
  ### Add middleware to your application

  The [`clerkMiddleware()`](/docs/references/express/overview#clerk-middleware) helper integrates Clerk authentication into your Express application. It can be registered globally or for specific routes. This following example registers it globally.
```

  ```ts
  import { getAuth } from '@clerk/express'

  app.get('/auth', (req, res) => {
    res.json({ auth: getAuth(req) })
  })
  ```

  ### Protect your routes using `requireAuth()`

  To protect your routes, use the [`requireAuth`](/docs/references/express/overview#require-auth) middleware. This middleware functions similarly to `clerkMiddleware()`, but also protects your routes by redirecting unauthenticated users to the sign-in page.

  In this example, if the user is authenticated, it retrieves the `userId` using the `getAuth()` helper and uses it with `clerkClient.users.getUser()` to fetch the current user's `User` object. If the user is not authenticated, they are redirected to the '/sign-in' route.

  ```ts {{ filename: 'index.ts', mark: [3, [7, 11], [13, 16]] }}
  import 'dotenv/config'
  import express from 'express'
  import { clerkClient, getAuth, requireAuth } from '@clerk/express'

  const app = express()

  app.get('/protected', requireAuth({ signInUrl: '/sign-in' }), (req, res) => {
    const { userId } = getAuth(req)
    const user = await clerkClient.users.getUser()
    return res.json({ user })
  })

  app.get('/sign-in', (req, res) => {

>     // Assuming you have a template engine installed

Reviewer 3: ```suggestion
    // Assuming you have a template engine installed and are using a Clerk javascript SDK on this page
```

<Steps>
  ### Install `@clerk/express`

  Clerk's [Express SDK](/docs/references/express/overview) has prebuilt components, React hooks, and helpers to make user authentication easier.

  To get started using Clerk with Express, add the SDK to your project:

  <CodeBlockTabs options={["npm", "yarn", "pnpm"]}>
    ```bash {{ filename: 'terminal' }}
    npm install @clerk/express
    ```

    ```bash {{ filename: 'terminal' }}
    yarn add @clerk/express
    ```

    ```bash {{ filename: 'terminal' }}
    pnpm add @clerk/express
    ```
  </CodeBlockTabs>

  ### Set your Clerk API keys

  <SignedIn>
    Add the following keys to your `.env.local` file. These keys can always be retrieved from the [API Keys](https://dashboard.clerk.com/last-active?path=api-keys) page of your Clerk Dashboard.
  </SignedIn>

  <SignedOut>
    1. Navigate to the Clerk Dashboard.
    1. In the navigation sidebar, select [API Keys](https://dashboard.clerk.com/last-active?path=api-keys).
    1. In the **Quick Copy** section, copy your Clerk publishable and secret key.
    1. Paste your keys into your `.env.local` file.

    The final result should resemble the following:
  </SignedOut>

>   ```sh {{ filename: '.env.local' }}

Reviewer 2: That's true. Updated!

  CLERK_PUBLISHABLE_KEY={{pub_key}}
  CLERK_SECRET_KEY={{secret}}
  ```

  This guide uses `dotenv` to load the environment variables. Install the package by running the following command:

  <CodeBlockTabs options={["npm", "yarn", "pnpm"]}>
    ```bash {{ filename: 'terminal' }}
    npm install dotenv

>     npm install -D @types/dotenv

Reviewer 3: I'm not sure how I feel about assuming typescript use here by default. Maybe a note below saying "if you are using typescript you will also need to install the external types" or something?

Reviewer 2: Updated and just removed the types installation for this one.

    ```

    ```bash {{ filename: 'terminal' }}
    yarn add dotenv
    yarn add -D @types/dotenv
    ```

    ```bash {{ filename: 'terminal' }}
    pnpm add dotenv
    pnpm add -D @types/dotenv
    ```
  </CodeBlockTabs>

  ### Add middleware to your application

  The [`clerkMiddleware()`](/docs/references/express/overview#clerk-middleware) helper integrates Clerk authentication into your Express application. It can be registered globally or for specific routes. The following example registers it globally.

  ```ts {{ filename: 'index.ts', mark: [3, 8] }}
  import 'dotenv/config'
  import express from 'express'
  import { clerkMiddleware } from '@clerk/express'

  const app = express()
  const PORT = 3000

  app.use(clerkMiddleware())

  app.listen(PORT, () => {
    console.log(`Example app listening at http://localhost:${PORT}`)
  })
  ```

  ### Accessing auth state using `getAuth()`

  The [`getAuth()`](/docs/references/express/overview#get-auth) helper retrieves authentication state from the request object.

>   [`clerkClient`](/docs/references/express/overview#clerk-client) is an instance of Clerk's JavaScript Backend SDK, which provides access to Clerk's Backend API resources and low-level authentication utilities for JavaScript environments.

Reviewer: because `clerkClient` is used in the code example, we should probably explain to the user what it is and what its doing. if that means the title of the section needs to change, thats true - but i wanted this copy added for the sake of the code example.

  The following example uses `getAuth()` to retrieve the `userId`, which is used to protect the route and is passed to `clerkClient.users.getUser()` to retrieve the current user's `User` object.

  > [!NOTE]

>   > If you would rather use middleware to protect routes, check out [`requireAuth`](/docs/references/express/overview#require-auth).

Reviewer 3: What are the tradeoffs here? Why would you want to use middleware vs doing it manually? I can't imagine, as a developer, coming to this guide with a pre-baked opinion on whether I wanted to do it like this or use middleware, without knowing what either of them are.

I also kind of feel like the default recommendation should be middleware, and this should be the "if you're trying to do something other than redirecting to sign in if the user isn't signed in and if they are, putting the auth state on the request object, you can do it custom like this" - but I'd bet the majority of people won't need to do it like that

Reviewer: requireAuth doesn't redirect to sign in, it returns an error.

the differences are, if you use requireAuth, it will return an error. you can also provide a custom response using error middleware.
and then if you use getAuth, again, you are providing a custom response to the unauthenticated user.

and you're saying, for the quickstart, we should show how to protect a route using requireAuth, and simply use getAuth to get the user's id for something like querying the backend api ?
